src/crimson/osd/osd_operations/snaptrim_event.cc

```diff
@@ -249,28 +249,6 @@ SnapTrimObjSubEvent::adjust_snaps(
 ) {
   head_obc->ssc->snapset.clone_snaps[coid.snap] =
     std::vector<snapid_t>(new_snaps.rbegin(), new_snaps.rend());
-
-  // we still do a 'modify' event on this object just to trigger a
-  // snapmapper.update ... :(
-  obc->obs.oi.prior_version = obc->obs.oi.version;
-  obc->obs.oi.version = osd_op_p.at_version;
-  ceph::bufferlist bl;
-  encode(obc->obs.oi,
-    bl,
-    pg->get_osdmap()->get_features(CEPH_ENTITY_TYPE_OSD, nullptr));
-  txn.setattr(
-    pg->get_collection_ref()->get_cid(),
-    ghobject_t{coid, ghobject_t::NO_GEN, shard_id_t::NO_SHARD},
-    OI_ATTR,
-    bl);
-  add_log_entry(
-    pg_log_entry_t::MODIFY,
-    coid,
-    obc->obs.oi.prior_version,
-    0,
-    osd_reqid_t(),
-    obc->obs.oi.mtime,
-    0);
   return OpsExecuter::snap_map_modify(
     coid, new_snaps, pg->snap_mapper, pg->osdriver, txn);
```

Comment:
I'm not actually sure this part is correct either -- this needs to be handled on the replicas as well, so it should be part of the log entry submission process.  Probably also worth double checking?

Reply:
Maybe we can add a new field in RepOp for only updating the in-memory snapmapper, instead of relying on pglog which is mainly used to record ondisk changes. What do you think?

Reply:
The snapmappers are ondisk state.  This mechanism already exists in classic -- we want to reuse it as it is part of the commit protocol.

Reply:
I can take care of that next week -- we can merge this simpler change in the meantime.

Reply:
https://tracker.ceph.com/issues/67057

Reply:
```
   t->update_snaps(
      coid,
      old_snaps,
      new_snaps);
```

in PrimaryLogPG::trim_object is actually a call to PGTransaction::update_snaps, which sets an internal map in the PGTransaction structure.  That map is then consulted in ReplicatedBackend.cc:generate_transaction:

```
  for (auto &&le: log_entries) {
    le.mark_unrollbackable();
    auto oiter = pgt->op_map.find(le.soid);
    if (oiter != pgt->op_map.end() && oiter->second.updated_snaps) {
      bufferlist bl(oiter->second.updated_snaps->second.size() * 8 + 8);
      encode(oiter->second.updated_snaps->second, bl);
      le.snaps.swap(bl);
      le.snaps.reassign_to_mempool(mempool::mempool_osd_pglog);
    }
  }
```

to populate the pg_log_entry_t::snaps entry on the log entry mutating the clone.  The correct behavior here would be to simply encode and set pg_log_entry_t::snaps for the clone entry generated.

Reply:
Actually updating the snap mapper should be done as part of (crimson) PG::log_operation.  It's actually if-def'd out:

```
#if 0
  if (transaction_applied) {
    //TODO:
    //update_snap_map(logv, t);
  }
  auto last = logv.rbegin();
  if (is_primary() && last != logv.rend()) {
    projected_log.skip_can_rollback_to_to_head();
    projected_log.trim(cct, last->version, nullptr, nullptr, nullptr);
  }
#endif
```

The code to consult in classic would be PrimaryLogPG::log_operation and PG::update_snap_map.  This is a better solution because PG::log_operation is invoked on both primary and replica at transaction generation time.

Reply:
Got it:-)

Reply:
Seems like a goo timing to revive https://github.com/ceph/ceph/pull/56832